src/axom/spin/UniformGrid.hpp

```diff
@@ -451,18 +702,214 @@ void UniformGrid<T, NDIMS>::insert(const BoxType& BB, const T& obj)
+  const auto offsets_view = outOffsets;
+  const auto counts_view = outCounts;
+#ifdef AXOM_USE_RAJA
+  using reduce_pol = typename axom::execution_space<ExecSpace>::reduce_policy;
+  RAJA::ReduceSum<reduce_pol, IndexType> totalCountReduce(0);
+  // Step 1: count number of candidate intersections for each point
+  for_all<ExecSpace>(
+    qsize,
+    AXOM_LAMBDA(IndexType i) {
+      counts_view[i] = gridQuery.countCandidates(queryObjs[i]);
+      totalCountReduce += counts_view[i];
+    });
+
+  // Step 2: exclusive scan for offsets in candidate array
+  using exec_policy = typename axom::execution_space<ExecSpace>::loop_policy;
+  RAJA::exclusive_scan<exec_policy>(RAJA::make_span(outCounts.data(), qsize),
+                                    RAJA::make_span(outOffsets.data(), qsize),
+                                    RAJA::operators::plus<IndexType> {});
+
+  axom::IndexType totalCount = totalCountReduce.get();
+
+  // Step 3: allocate memory for all candidates
+  axom::Array<IndexType> queryIndex(totalCount,
+                                    totalCount,
+                                    this->getAllocatorID());
+  outCandidates =
+    axom::Array<IndexType>(totalCount, totalCount, this->getAllocatorID());
+  const auto query_idx_view = queryIndex.view();
+  const auto candidates_view = outCandidates.view();
+
+  // Step 4: fill candidate array for each query box
+  for_all<ExecSpace>(
+    qsize,
+    AXOM_LAMBDA(IndexType i) {
+      int startIdx = offsets_view[i];
+      int currCount = 0;
+      auto onCandidate = [&](int candidateIdx) -> bool {
+        candidates_view[startIdx] = candidateIdx;
+        query_idx_view[startIdx] = i;
+        currCount++;
+        startIdx++;
+        return currCount >= counts_view[i];
+      };
+      gridQuery.visitCandidates(queryObjs[i], onCandidate);
+    });
+
+  // Step 5: Sort our resulting candidates for each query object.
+  // This brings non-unique candidate intersections together.
+  if(axom::execution_space<ExecSpace>::onDevice())
+  {
+    // On the GPU, we first sort pairs by candidate index, then stable sort by
+    // the query index.
+    RAJA::sort_pairs<exec_policy>(
+      RAJA::make_span(outCandidates.data(), totalCount),
+      RAJA::make_span(queryIndex.data(), totalCount));
+    RAJA::stable_sort_pairs<exec_policy>(
+      RAJA::make_span(queryIndex.data(), totalCount),
+      RAJA::make_span(outCandidates.data(), totalCount));
+  }
+  else
+  {
+    // On the CPU, just sort each subrange for each query object.
+    for_all<ExecSpace>(
+      qsize,
+      AXOM_LAMBDA(IndexType i) {
+        int startIdx = offsets_view[i];
+        int count = counts_view[i];
+        if(count > 0)
+        {
+  #ifndef __CUDA_ARCH__
+          std::sort(candidates_view.begin() + startIdx,
+                    candidates_view.begin() + startIdx + count);
+  #endif
+        }
+      });
+  }
+
+  // Step 6: Count and flag unique intersection pairs, in order to map them
```

Comment:
Long-shot: Does RAJA have a direct way to sort and unique-ify an array?

Reply:
I don't think there is; in our use case it's complicated by the fact that we only want to unique-ify within individual result slices; not sure if we can really express that generically in RAJA.